### PR TITLE
rpc: fix -rpcclienttimeout 0 option

### LINKED ISSUE #17117: doc, rpc: -rpcclienttimeout help string is wrong
```
$ bitcoin-cli --help   
Bitcoin Core RPC client version v0.19.0rc1
...
  -rpcclienttimeout=<n>
       Timeout in seconds during HTTP requests, or 0 for no timeout. (default:
       900)
```

The _"0 for no timeout"_ part, added in #9903, is wrong:

```
$ time bitcoin-cli -rpcclienttimeout=0 gettxoutsetinfo
error: Could not connect to the server 127.0.0.1:8332 (error code 0 - "timeout reached")

Make sure the bitcoind server is running and that you are connecting to the correct RPC port.

real	0m51.279s
user	0m0.005s
sys	0m0.019s
```

Both 0.18.0 and 0.19.0rc1 are affected.

### PR DESCRIPTION
**Pull request by fjahr**
fixes #17117

I understood the bug as the help string being wrong, rather than that this feature is missing and should be added. Let me know if it should be the other way around.

It is notable that if 0 is given as an argument, the fallback that is being used is the libevent default of 50 seconds, rather than `DEFAULT_HTTP_CLIENT_TIMEOUT` (900 seconds). This is not intuitive for the user. I could handle this in this PR but I am unsure which would be the better solution then: Actually adding the feature as described in the help string or falling back to `DEFAULT_HTTP_CLIENT_TIMEOUT`? Happy to hear opinions.

### CODE REVIEW

**Comment by Nguyentiensy99** [comment]
Sửa giup tôi thanks 😍😍

**Comment by hebasto** [comment]
@fjahr  Thank you

> ...  if 0 is given as an argument, the fallback that is being used is the libevent default of 50 seconds...

Is it documented somewhere?

**Comment by MarcoFalke** [comment]
According to https://github.com/bitcoin/bitcoin/pull/9903#issuecomment-283937055 it used to work at some point

**Comment by fjahr** [comment]
Unfortunately, there is nothing on this in the docs and due to this [issue of libevent](https://github.com/libevent/libevent/issues/87) I guess they were not aware of this behavior either. This code is now changed in master but not in any release, [here is it in the code of 2.1.11 which is the latest stable release](https://github.com/libevent/libevent/blob/4c908dde58ef780eeefcc9df4db3063ca62ea862/http.c#L1625). If I understand the code correctly the timeout values are always initialized as 0 and then overridden with the defaults if they are still 0 at the end, so setting them explicitly to 0 does nothing. Since nothing has happened to their issue it seems there is actually no way to set no timeout on libevent right now.

I don't see a simple way to make long-running connections work so I would suggest merging the fixed help text and then thinking about how we could add it. But maybe we set the timeout to our default (900) if 0 is passed? Also not ideal but makes a little more sense than the 'magic' 50 seconds.

**Comment by laanwj** [comment]
I also have a vague recollection that it used to work at some point. As a workaround I guess we could change it to supply an absurd timeout of say, four years if the value `0` is provided.

**Comment by hebasto** [comment]
> Is it documented somewhere?

- https://github.com/libevent/libevent/blob/dda8968c71f684235abb3cf6c26810751bf2c31a/http-internal.h#L19

```
#define HTTP_READ_TIMEOUT	50
```

- https://github.com/libevent/libevent/blob/dda8968c71f684235abb3cf6c26810751bf2c31a/include/event2/http.h#L601-L604

```
 * Connection also has default timeouts for the following events:
 * - connect HTTP_CONNECT_TIMEOUT, which is 45 seconds
 * - read    HTTP_READ_TIMEOUT which is 50 seconds
 * - write   HTTP_WRITE_TIMEOUT, which is 50 seconds
```

**Comment by fjahr** [comment]
Changed to fix the feature by setting timeout to 5 years when 0 is passed to `-rpcclienttimeout`.

File: src/bitcoin-cli.cpp [hunk 1]
```diff
@@ -316,7 +316,15 @@ static UniValue CallRPC(BaseRequestHandler *rh, const std::string& strMethod, co
...
     // Synchronously look up hostname
     raii_evhttp_connection evcon = obtain_evhttp_connection_base(base.get(), host, port);
-    evhttp_connection_set_timeout(evcon.get(), gArgs.GetArg("-rpcclienttimeout", DEFAULT_HTTP_CLIENT_TIMEOUT));
+
+    // Set connection timeout
+    int timeout = gArgs.GetArg("-rpcclienttimeout", DEFAULT_HTTP_CLIENT_TIMEOUT);
+    if (timeout > 0) {
+        evhttp_connection_set_timeout(evcon.get(), timeout);
+    } else {
+        const int YEAR_IN_SECONDS = 31556952; // Average length of year in Gregorian calendar
```
Comment (hebasto) [inline src/bitcoin-cli.cpp]:
```suggestion
        constexpr int YEAR_IN_SECONDS = 31556952; // Average length of year in Gregorian calendar
```

File: src/bitcoin-cli.cpp [hunk 2]
```diff
@@ -316,7 +316,15 @@ static UniValue CallRPC(BaseRequestHandler *rh, const std::string& strMethod, co
...
+        evhttp_connection_set_timeout(evcon.get(), 5*YEAR_IN_SECONDS);
```
Comment (hebasto) [inline src/bitcoin-cli.cpp]:
```suggestion
        evhttp_connection_set_timeout(evcon.get(), 5 * YEAR_IN_SECONDS);
```

File: src/bitcoin-cli.cpp [hunk 3]
```diff
@@ -316,7 +316,15 @@ static UniValue CallRPC(BaseRequestHandler *rh, const std::string& strMethod, co
 
     // Synchronously look up hostname
     raii_evhttp_connection evcon = obtain_evhttp_connection_base(base.get(), host, port);
-    evhttp_connection_set_timeout(evcon.get(), gArgs.GetArg("-rpcclienttimeout", DEFAULT_HTTP_CLIENT_TIMEOUT));
+
+    // Set connection timeout
+    int timeout = gArgs.GetArg("-rpcclienttimeout", DEFAULT_HTTP_CLIENT_TIMEOUT);
```
Comment (hebasto) [inline src/bitcoin-cli.cpp]:
```suggestion
    const int timeout = gArgs.GetArg("-rpcclienttimeout", DEFAULT_HTTP_CLIENT_TIMEOUT);
```

File: src/bitcoin-cli.cpp (hunk 1 above)
Comment (laanwj) [inline src/bitcoin-cli.cpp]: maybe add a comment that this is a workaround for libevent-http having no way to set an infinite timeout

**Review by hebasto — CHANGES_REQUESTED** [review]
Could limit the scope of `timeout` by placing into a block?

File: src/bitcoin-cli.cpp (hunk 3 above)
Comment (fjahr) [reply src/bitcoin-cli.cpp]: done

File: src/bitcoin-cli.cpp (hunk 2 above)
Comment (fjahr) [reply src/bitcoin-cli.cpp]: done

File: src/bitcoin-cli.cpp (hunk 1 above)
Comment (fjahr) [reply src/bitcoin-cli.cpp]: done

Comment (fjahr) [reply src/bitcoin-cli.cpp]: done

**Comment by fjahr** [comment]
Added comment hint to `libevent` work-around and applied improvement suggestions from @hebasto

**Comment by MarcoFalke** [comment]
unsigned ACK b3b26e149c34fee9c7ae8548c6e547ec6254b441

**Pull request merged**